Rubberduck.Parsing/Rewriter/MemberAttributeRecoverer.cs

```diff
@@ -0,0 +1,233 @@
+            }
+
+            foreach (var declaration in declarationsWithAttributesToRecover)
+            {
+                RecoverAttributes(rewriteSession, declaration, attributesByMember[declaration.IdentifierName]);
+            }
+        }
+
+        private static ICollection<string> MembersWithoutDeclarations(HashSet<string> membersWithAttributesToRecover, IEnumerable<Declaration> declarationsWithAttributesToRecover)
+        {
+            var membersWithoutDeclarations = membersWithAttributesToRecover.ToHashSet();
+            membersWithoutDeclarations.ExceptWith(declarationsWithAttributesToRecover.Select(decl => decl.IdentifierName));
+            return membersWithoutDeclarations;
+        }
+
+        private void LogFailureToRecoverAllAttributes(QualifiedModuleName module, IEnumerable<string> membersWithoutDeclarations)
+        {
+            _logger.Warn("Could not recover the attributes for all members because one or more members could no longer be found.");
+            foreach (var member in membersWithoutDeclarations)
+            {
+                _logger.Trace($"Could not recover the attributes for member {member} in module {module} because a member of that name exists no longer.");
+            }
+        }
+
+        private void RecoverAttributes(IRewriteSession rewriteSession, Declaration declaration, IEnumerable<AttributeNode> attributes)
+        {
+            foreach (var attribute in attributes)
+            {
+                if (!declaration.Attributes.Contains(attribute))
+                {
+                    _attributesUpdater.AddAttribute(rewriteSession, declaration, attribute.Name, attribute.Values);
+                }
+            }
+        }
+
+        private void EndTheCurrentParse(CancellationToken currentCancellationToken)
+        {
+            currentCancellationToken.ThrowIfCancellationRequested();
+        }
+
+        public void RecoverCurrentMemberAttributesAfterNextParse(IEnumerable<QualifiedModuleName> modules)
+        {
+            var declarationsByModule = MemberDeclarationsByModule(modules);
+            RecoverCurrentMemberAttributesAfterNextParse(declarationsByModule);
+        }
+
+        private IDictionary<QualifiedModuleName, IEnumerable<Declaration>> MemberDeclarationsByModule(IEnumerable<QualifiedModuleName> modules)
+        {
+            var declarationFinder = _declarationFinderProvider.DeclarationFinder;
+            var declarationsByModule = modules.ToDictionary(
+                module => module, 
+                module => declarationFinder.Members(module)
+                    .Where(decl => !decl.DeclarationType.HasFlag(DeclarationType.Module)
+                                   && decl.ParentScopeDeclaration.DeclarationType.HasFlag(DeclarationType.Module)));
+            return declarationsByModule;
+        }
+
+        public void Dispose()
```

Comment:
Can we use the "MS sanctioned" Dispose pattern here? These spam the build output with warnings.

Reply:
Will the warnings also go away if I seal the class? There is no intent to introduce subclasses and there are no unmanaged recourse. So, the full pattern makes little sense here.

Reply:
Sealing the class should work.